### PR TITLE
Feature/gridedit 777 casulli refinement

### PR DESCRIPTION
**Pull request by BillSenior**

### CODE REVIEW

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+            }
+
+            if (mesh.m_edges[link[j]].second == i && nodeMask[newNodes[link[j]][3]] == -1)
+            {
+                node[j] = newNodes[link[j]][3];
+            }
+
+            //
+
+            if (mesh.m_edges[link[j]].first == i && nodeMask[newNodes[link[j]][0]] == 1236)
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliRefinement.cpp:412]: Replace numbers such as 1236 with enumerations

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+meshkernel::UInt meshkernel::CasulliRefinement::IsLeftRight(const Mesh2D& mesh, const UInt elementId, const UInt edgeId)
+{
+    UInt isLeftRight = constants::missing::uintValue;
+    UInt edgeIndex = constants::missing::uintValue;
+    UInt nextEdgeIndex = constants::missing::uintValue;
+    UInt endNodeIndex = mesh.m_edges[edgeId].second;
+
+    for (UInt i = 0; i < mesh.m_facesEdges[elementId].size(); ++i)
+    {
+        UInt faceEdgeId = mesh.m_facesEdges[elementId][i];
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliRefinement.cpp]: Use const when you can

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+    }
+    else if (edgeIndex == nextEdgeIndex + 1 || edgeIndex + mesh.m_numFacesNodes[elementId] == nextEdgeIndex + 1)
+    {
+        isLeftRight = 1;
+    }
+
+    return isLeftRight;
+}
+
+meshkernel::UInt meshkernel::CasulliRefinement::FindCommon(const Mesh2D& mesh, const UInt edge1, const UInt edge2)
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliRefinement.cpp]: Should this be moved to Mesh2D?

**Review by lucacarniato — CHANGES_REQUESTED** [review]

File: libs/MeshKernel/include/MeshKernel/CasulliRefinement.hpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    private:
+        // @brief Initial size of the edge array
+        static constexpr UInt InitialEdgeArraySize = 100;
+
+        /// @brief The maximum number of nodes that a newly created element can have.
+        static constexpr UInt MaximumNumberOfNodesInNewlyCreatedElements = 4;
+
+        /// @brief Simple bounded array of 4 UInt values
+        /// @typedef LinkNodes
+        using LinkNodes = std::array<UInt, 4>;
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CasulliRefinement.hpp:57]: the 4 is not for quadrilaterals. The comment in the fortran is

newnodes   ! four new nodes per existing link

File: libs/MeshKernel/src/CasulliRefinement.cpp (hunk 3 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp]: Its been moved already, together with a couple of other functions.

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+            // two new and opposing nodes: new diagonal link connects the new nodes
+            if (nodeMask[node1] < 0 && nodeMask[node2] == 0 && nodeMask[node3] == 0 && nodeMask[node4] == 1)
+            {
+                mesh.ConnectNodes(node1, newIndex[nextNextIndex]);
+                break;
+            }
+        }
+    }
+
+    std::vector<UInt> link(InitialEdgeArraySize);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp:337]: Changed (to newEdges)

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+
+    ComputeNewNodes(mesh, newNodes, nodeMask);
+    LinkNewNodes(mesh, newNodes, numNodes, numEdges, numFaces, nodeMask);
+    Administrate(mesh, numNodes, nodeMask);
+}
+
+std::vector<int> meshkernel::CasulliRefinement::InitialiseNodeMask(const Mesh2D& mesh, const Polygons& polygon)
+{
+    // Need a better initial size
+    std::vector<int> nodeMask(10 * mesh.GetNumNodes(), 0);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp]: Changed the array element to the (newly added) enum . The enum type is defined to be a char, so should be a quarter of the memory used when compared to int.

File: libs/MeshKernel/src/CasulliRefinement.cpp (hunk 1 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp]: All node mask numbers have been replaced with enum values

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+    {
+        UInt node1 = mesh.m_edges[i].first;
+        UInt node2 = mesh.m_edges[i].second;
+
+        if (mesh.m_edgesNumFaces[i] == 1)
+        {
+
+            if (nodeMask[node1] != 0)
+            {
+                nodeMask[node1] = 1234;
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp]: All node mask values have been replaced with enum values

File: libs/MeshKernel/include/MeshKernel/Mesh2D.hpp [hunk 8]
```diff
@@ -327,6 +327,43 @@ namespace meshkernel
...
+        /// @param [in] sharedFaces List of faces that share the nodeIndex as a common node
+        /// @param [in, out] connectedNodes List of nodes that are in the patch of shared faces
+        /// @param [out] faceNodeMapping Mapping from node index to the position in connectedNodes list.
+        void FindNodesSharedByFaces(UInt nodeIndex, const std::vector<UInt>& sharedFaces, std::vector<UInt>& connectedNodes, std::vector<std::vector<UInt>>& faceNodeMapping) const;
+
+        /// @brief Determine if the node is at the start or end of the edge.
+        ///
+        /// Returns 0 when the node is at the start of the edge, 1 when it is at the end
+        /// and the null value when the edge is not connected to the node.
+        UInt IsStartOrEnd(const UInt edgeId, const UInt nodeId) const;
```
Comment (lucacarniato) [inline libs/MeshKernel/include/MeshKernel/Mesh2D.hpp:354]: in entities there is a function OtherNodeOfEdge, shall we move OtherNodeOfEdge also in mesh2d class?

File: libs/MeshKernel/include/MeshKernel/Mesh2D.hpp [hunk 9]
```diff
@@ -327,6 +327,43 @@ namespace meshkernel
...
+        /// @brief Determine if the element lies on the left or right side of the edge
+        ///
+        /// Returns 0 when the element is on the left and 1 when it is on the right.
+        /// If one or other edge is not connected to the element then a null value will be returned.
+        UInt IsLeftOrRight(const UInt elementId, const UInt edgeId) const;
+
+        /// @brief Find the id of the element that is common to both edges.
+        ///
+        /// If no such element can be found then the null value will be returned.
+        UInt FindCommonElement(const UInt edge1, const UInt edge2) const;
```
Comment (lucacarniato) [inline libs/MeshKernel/include/MeshKernel/Mesh2D.hpp:365]: FindCommonFace, for keeping the implementation consistent

**Review by lucacarniato — APPROVED** [review]

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 10]
```diff
@@ -0,0 +1,621 @@
...
+        {
+            UInt edge1 = mesh.m_nodesEdges[i][j];
+
+            if (mesh.m_edgesNumFaces[edge1] != 1)
+            {
+                continue;
+            }
+
+            UInt elementId = mesh.m_edgesFaces[edge1][0];
+            UInt nodeCount = mesh.m_numFacesNodes[elementId];
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/src/CasulliRefinement.cpp:76]: const

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 11]
```diff
@@ -0,0 +1,621 @@
...
+#include "MeshKernel/CasulliRefinement.hpp"
+#include "MeshKernel/Exceptions.hpp"
+
+void meshkernel::CasulliRefinement::Compute(Mesh2D& mesh)
+{
+    Polygons emptyPolygon;
+    Compute(mesh, emptyPolygon);
+}
+
+void meshkernel::CasulliRefinement::Compute(Mesh2D& mesh, const Polygons& polygon)
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/src/CasulliRefinement.cpp:12]: Too long. Can i t be broken down?

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 12]
```diff
@@ -0,0 +1,621 @@
...
+    if (sharedFaces.empty())
+    {
+        return;
+    }
+
+    mesh.GetConnectingNodes(currentNode, connectedNodes);
+    mesh.FindNodesSharedByFaces(currentNode, sharedFaces, connectedNodes, faceNodeMapping);
+}
+
+void meshkernel::CasulliRefinement::ConnectNewNodes(Mesh2D& mesh, const std::vector<EdgeNodes>& newNodes, const UInt numNodes, const UInt numEdges, const UInt numFaces, std::vector<NodeMask>& nodeMask)
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/src/CasulliRefinement.cpp:205]: Too long. Can i t be broken down?

File: libs/MeshKernel/src/CasulliRefinement.cpp [hunk 13]
```diff
@@ -0,0 +1,621 @@
...
+            else
+            {
+                mesh.ConnectNodes(i, newNodes[edgeId][1]);
+                mesh.ConnectNodes(i, newNodes[edgeId][3]);
+            }
+        }
+    }
+}
+
+void meshkernel::CasulliRefinement::ComputeNewNodes(Mesh2D& mesh, std::vector<EdgeNodes>& newNodes, std::vector<NodeMask>& nodeMask)
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/src/CasulliRefinement.cpp:460]: Too long. Can i t be broken down?

File: libs/MeshKernel/src/CasulliRefinement.cpp (hunk 12 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp]: Refactored

File: libs/MeshKernel/src/CasulliRefinement.cpp (hunk 13 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp]: Refactored

File: libs/MeshKernel/src/CasulliRefinement.cpp (hunk 11 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliRefinement.cpp:12]: Refactored

**Pull request merged**